clients/mobile/src/setup/notification/NotificationProvider.tsx

```diff
@@ -1,86 +1,94 @@
-          }`
-        )
+    if (!listenerAddedRef.current) {
+      const unsubscribe = navigationRef.addListener('state', () => {
+        if (navigationRef.isReady() && !isReady) setIsReady(true)
+      })
+      listenerAddedRef.current = true
+      return () => {
+        unsubscribe()
+        listenerAddedRef.current = false
       }
     }
+  }, [navigationRef, isReady])
 
-    void checkInitialNotification()
+  useEffect(() => {
+    if (!isReady) return
 
-    const foregroundListener = Notifications.addNotificationReceivedListener(notification => {
-      const data = parseJsonData<NotificationData>(notification.request.content.data.payload)
-      if (data !== null) setNotificationData(data)
-    })
+    Notifications.getLastNotificationResponseAsync().then(response => {
+      if (response === null || !isNotificationValid(response)) return
 
-    const responseListener = Notifications.addNotificationResponseReceivedListener(response => {
-      const data = parseJsonData<NotificationData>(response.notification.request.content.data.payload)
-      if (isNotificationValid(response, isMounted) && data !== null) {
-        setNotificationData(data)
-        navigation.navigate(SCREENS.BLOOD_REQUEST_PREVIEW, { notificationData: data })
+      const identifier = response.notification.request.identifier
+      if (lastProcessedNotification.current === identifier) return
+
+      lastProcessedNotification.current = identifier
+      const data = parseJsonData(response.notification.request.content.data.payload)
+      setNotificationData(data as NotificationData)
+      handleNotificationNavigation(response)
+    }).catch(() => { })
+
+    const responseListener = Notifications.addNotificationResponseReceivedListener((response) => {
+      if (isNotificationValid(response)) {
+        const data = parseJsonData(response.notification.request.content.data.payload)
+        setNotificationData(data as NotificationData)
+        handleNotificationNavigation(response)
       }
     })
 
-    return () => {
-      isMounted = false
-      foregroundListener.remove()
-      responseListener.remove()
+    return () => { responseListener.remove() }
+  }, [navigationRef, isReady])
+
+  const isNotificationValid = (response: Notifications.NotificationResponse | null): boolean => {
+    return (
+      Object.keys(response?.notification.request.content.data.payload ?? {}).length > 0 &&
+      response?.notification.request.identifier !== null
+    )
+  }
+
+  const handleNotificationNavigation = (response: Notifications.NotificationResponse | null) => {
+    if (response === null) return
+
+    const { type } = response.notification.request.content.data
+    const mapping = SCREEN_FOR_NOTIFICATION[type]
+
+    if (mapping !== undefined) {
+      const { screen, getParams } = mapping
+      const params = getParams !== undefined ? getParams(parseJsonData<Record<string, unknown>>(response.notification.request.content.data.payload)) : undefined
```

Comment:
Line length is too long

Reply:
Are there any rules for our project about how much a line should be?

Reply:
not yet, but it should must be less than 100